File: src/security/injection-defense.ts
export class PromptInjectionDefense {
  private static readonly INJECTION_PATTERNS = [
    /ignore (all )?previous instructions/i,
    /system prompt/i,
    /you are now/i,
    /give (me|yourself) admin/i,
    /bypass permission/i,
    /eval\s*\(/i,
    /child_process/i,
    /execSync/i,
  ];

  public static sanitizeInput(input: string): string {
    // Strip control characters while preserving utf-8 natural text
    return input.replace(/[\u0000-\u0008\u000B\u000C\u000E-\u001F\u007F-\u009F]/g, '').trim();
  }

  public static detectInjectionAttempt(input: string): boolean {
    return this.INJECTION_PATTERNS.some((pattern) => pattern.test(input));
  }
}

File: bot-py/src/security/injection_defense.py
import re


class InjectionDefense:
    DANGEROUS_SYSTEM_PATTERNS = [
        re.compile(r"ignore (?:all|previous|\s)+ (?:instructions|directives)", re.IGNORECASE),
        re.compile(r"you are now in (?:developer|dan|admin) mode", re.IGNORECASE),
        re.compile(r"grant (?:me|yourself) administrator", re.IGNORECASE),
        re.compile(r"system prompt", re.IGNORECASE),
        re.compile(r"eval\s*\(", re.IGNORECASE),
        re.compile(r"exec\s*\(", re.IGNORECASE),
        re.compile(r"subprocess", re.IGNORECASE),
        re.compile(r"child_process", re.IGNORECASE),
    ]

    @staticmethod
    def sanitize_external_text(text: str) -> str:
        # Strip control characters
        cleaned = re.sub(r"[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]", "", text)
        return cleaned.strip()

    @classmethod
    def detect_prompt_injection(cls, prompt: str) -> bool:
        return any(pattern.search(prompt) for pattern in cls.DANGEROUS_SYSTEM_PATTERNS)
